evmcore/apply_fake_genesis.go

```diff
@@ -93,12 +94,123 @@ func MustApplyFakeGenesis(statedb *state.StateDB, time inter.Timestamp, balances
+	var key82, _ = crypto.ToECDSA(hexutil.MustDecode("0x95f963b910d735ffd9dfd784ca683cc790aa40acdc5fa9d27fe8e84b6900528d"))
+	var key83, _ = crypto.ToECDSA(hexutil.MustDecode("0xe8fa32934f5e6c895f1586114cd3c84caf04f01efab47bcd8bc2efbd743e6dfb"))
+	var key84, _ = crypto.ToECDSA(hexutil.MustDecode("0x030076ec9dd721e30c743bea5f0019e80428a97fe900b7245b968c6a6f774313"))
+	var key85, _ = crypto.ToECDSA(hexutil.MustDecode("0xb7001594545d584baab7d6b056d3a39c325db2b450787329ed916dd26fa32260"))
+	var key86, _ = crypto.ToECDSA(hexutil.MustDecode("0xd2a748ae6cfa9156c94754320b22b35b295dfe779887098386b2cea72f4d0dd2"))
+	var key87, _ = crypto.ToECDSA(hexutil.MustDecode("0x258bb5bc5a87c5aebdadb7f83f39029242f0650bc52fc4cc482c89412532db2b"))
+	var key88, _ = crypto.ToECDSA(hexutil.MustDecode("0x20ecd5168010ed59a184fd5b5ac528b9fff70ed103ea58c6f35f0137854969b8"))
+	var key89, _ = crypto.ToECDSA(hexutil.MustDecode("0x1a517bb0a54425b29b032d4cc28423224963cda64b50ef9731b429660c8da129"))
+	var key90, _ = crypto.ToECDSA(hexutil.MustDecode("0xee61ccd710d6dabb81849a1ca2a9369c5484b65517ea80c7772135c55aa9f147"))
+	var key91, _ = crypto.ToECDSA(hexutil.MustDecode("0xe906f9f17d6511279053a95feca4a55ef59f97f2596790163ade48067d20238f"))
+	var key92, _ = crypto.ToECDSA(hexutil.MustDecode("0x3c8cd94010f44ac08efaae8a31e726bb4fc95564ee7c2e03e7a43ee43f31d6ed"))
+	var key93, _ = crypto.ToECDSA(hexutil.MustDecode("0x58ed1a9bde738f0a089ae0c7e18bf77d2489e75f282c1af6e16e0b86fc30c41e"))
+	var key94, _ = crypto.ToECDSA(hexutil.MustDecode("0x0be230f443fdc623254292438f0c0f86ba0b799c89d15c9383aa3a64d99628e7"))
+	var key95, _ = crypto.ToECDSA(hexutil.MustDecode("0x26a53f1c2b8fff8cfc6ca691777d1c84eb6bef60a3b3c26233c1b12faf653584"))
+	var key96, _ = crypto.ToECDSA(hexutil.MustDecode("0x79a6be35f01db3d6a2b659b49f04505a0ce1abf3770494a7a24a83642ae8a675"))
+	var key97, _ = crypto.ToECDSA(hexutil.MustDecode("0x54cd7ec556aeeb6f538fe1b7523a86425e520e288be516c95909582e79012bc3"))
+	var key98, _ = crypto.ToECDSA(hexutil.MustDecode("0x675e4d1f766213db0791a274e748e858d091460fb4a9222e4b75380f8ecabcdc"))
+	var key99, _ = crypto.ToECDSA(hexutil.MustDecode("0x4203c438d4e94bf4a595794b5f5c2882f959face730abb7a7b8acb462c8e138d"))
+
+	var keys = [100]*ecdsa.PrivateKey{
+		key00, key01, key02, key03, key04, key05, key06, key07, key08, key09,
+		key10, key11, key12, key13, key14, key15, key16, key17, key18, key19,
+		key20, key21, key22, key23, key24, key25, key26, key27, key28, key29,
+		key30, key31, key32, key33, key34, key35, key36, key37, key38, key39,
+		key40, key41, key42, key43, key44, key45, key46, key47, key48, key49,
+		key50, key51, key52, key53, key54, key55, key56, key57, key58, key59,
+		key60, key61, key62, key63, key64, key65, key66, key67, key68, key69,
+		key70, key71, key72, key73, key74, key75, key76, key77, key78, key79,
+		key80, key81, key82, key83, key84, key85, key86, key87, key88, key89,
+		key90, key91, key92, key93, key94, key95, key96, key97, key98, key99,
+	}
```

Comment:
I would change the type of `n` to `uint` to satisfy your condition at L211 as `n` can not be negative. I think, there is an idiomatic way to declare multiple variables inside `FakeKey` like that:
```
var (
     key = ...
     keys = [100]*ecdsa.PrivateKey{
)
```

Reply:
> I think, there is an idiomatic way to declare multiple variables inside FakeKey like that:

Okay, I write code to generate this code segment so didn't notice this

> I would change the type of n to uint 

`idx.ValidatorID` type actually is `uint32` to I changed the type of `n` to `uint32` as well

Reply:
Wondering if these will work:

var seed = []byte("163f5f0f9a621d72fedd85ffca3d08d131ab4e812181e0d30ffd1c885d20aac73144c0aa4ced56dc15c79b045bc5559a5ac9");

.......

// FakeKey gets n-th fake private key.
func FakeKey(n uint32) *ecdsa.PrivateKey {
	if n == 0 || n > 100 {
			panic(errors.New("validator num is out of range"))
	}

	// shift left n characters of the seed to the end
	var shifted = append(seed[n:], seed[:n]...)
	// and get the first 64 bytes
	var key = "0x" + shifted[:64];

	var key, err = crypto.ToECDSA(hexutil.MustDecode(key))

 	if err != nil {
 		panic(err)
 	}

	return key
}

Reply:
Yes, I think it might work but we want to keep the **fakekey** as same as when running below code on old version of golang. So we pre-generated all the first 100 keys from old version of golang and hardcode into the source like that.

```
	reader := rand.New(rand.NewSource(int64(n)))

	key, err := ecdsa.GenerateKey(crypto.S256(), reader)
```